### PR TITLE
add bundle.go to verify a bundle

### PR DESCRIPTION
**Pull request by liangchenye**
Signed-off-by: liangchenye liangchenye@huawei.com

### CODE REVIEW

File: bundle.go [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+	for index := 0; index < len(mps); index++ {
+		mountPath := path.Join(rootfs, mps[index].Path)
+		if fi, err := os.Stat(mountPath); err != nil {
+			logrus.Fatal(err)
+		} else if !fi.IsDir() {
+			logrus.Fatalf("The mountPoint %s is not a valid directory", mountPath)
+		}
+
+		if _, ok := rmps[mps[index].Name]; !ok {
+			logrus.Fatalf("%s in config/mount is not exist in runtime/mount", mps[index].Name)
```
Comment (mrunalp) [inline bundle.go]: s/is/does

File: bundle.go [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+		}
+		if !found {
+			logrus.Fatalf("The mount %s is missing.", paths[pIndex])
+		}
+	}
+
+	for index := 0; index < len(spec.Linux.Capabilities); index++ {
+		capability := spec.Linux.Capabilities[index]
+		if !capValid(capability) {
+			logrus.Fatalf("%s is not valid, please `man capabilities`", spec.Linux.Capabilities[index])
```
Comment (mrunalp) [inline bundle.go]: please see

File: bundle.go [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+		}
+	}
+}
+
+//Linux only
+func CheckLinuxRuntime(runtime specs.LinuxRuntime, rootfs string) {
+	if len(runtime.UIDMappings) > 5 {
+		logrus.Fatalf("The UID mapping is limited to 5.")
```
Comment (mrunalp) [inline bundle.go]: Only 5 UID mappings are allowed (linux kernel restriction)

File: bundle.go [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+	}
+	if len(runtime.GIDMappings) > 5 {
+		logrus.Fatalf("The GID mapping is limited to 5.")
```
Comment (mrunalp) [inline bundle.go]: Same here

File: bundle.go [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+	}
+
+	for index := 0; index < len(runtime.Rlimits); index++ {
+		if !rlimitValid(runtime.Rlimits[index].Type) {
+			logrus.Fatalf("The Rlimit %s is invalid.", runtime.Rlimits[index])
```
Comment (mrunalp) [inline bundle.go]: Take out 'The'

File: bundle.go [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+		}
+	}
+
+	for index := 0; index < len(runtime.Namespaces); index++ {
+		if !namespaceValid(runtime.Namespaces[index]) {
+			logrus.Fatalf("The Namespace %s is invalid.", runtime.Namespaces[index])
```
Comment (mrunalp) [inline bundle.go]: Remove The

**Comment by liangchenye** [comment]
Updated, Thanks!

File: bundle.go [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+		defer sf.Close()
+
+		//TODO: Using protobuf to decode spec, so the 'required' config will be checked automaticly
+		var spec specs.LinuxSpec
+		if err = json.NewDecoder(sf).Decode(&spec); err != nil {
+			logrus.Fatal(err)
+		} else {
+			if spec.Platform.OS != "linux" {
+				logrus.Fatalf("The operation system '%s' of the bundle is not support yet.", spec.Platform.OS)
```
Comment (mrunalp) [inline bundle.go]: Remove The

File: bundle.go [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+	}
+}
+
+func CheckMountPoints(mps []specs.MountPoint, rmps map[string]specs.Mount, rootfs string) {
+	for index := 0; index < len(mps); index++ {
+		mountPath := path.Join(rootfs, mps[index].Path)
+		if fi, err := os.Stat(mountPath); err != nil {
+			logrus.Fatal(err)
+		} else if !fi.IsDir() {
+			logrus.Fatalf("The mountPoint %s is not a valid directory", mountPath)
```
Comment (mrunalp) [inline bundle.go]: Remove The

Comment (mrunalp) [inline bundle.go]: Mount point

File: bundle.go [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+		}
+		if !found {
+			logrus.Fatalf("The mount %s is missing.", paths[pIndex])
+		}
+	}
+
+	for index := 0; index < len(spec.Linux.Capabilities); index++ {
+		capability := spec.Linux.Capabilities[index]
+		if !capValid(capability) {
+			logrus.Fatalf("%s is not valid, please see `man capabilities`", spec.Linux.Capabilities[index])
```
Comment (mrunalp) [inline bundle.go]: man capabilities(7)

File: bundle.go [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+	for index := 0; index < len(devices); index++ {
+		found := false
+		for dIndex := 0; dIndex < len(runtime.Devices); dIndex++ {
+			if runtime.Devices[dIndex].Path == devices[index] {
+				found = true
+				break
+			}
+		}
+		if found == false {
+			logrus.Fatalf("The required device %s is missing.", devices[index])
```
Comment (mrunalp) [inline bundle.go]: Remove The

File: bundle.go [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+		}
+	}
+
+	for index := 0; index < len(runtime.Devices); index++ {
+		if !deviceValid(runtime.Devices[index]) {
+			logrus.Fatalf("The device %s is invalid.", runtime.Devices[index].Path)
```
Comment (mrunalp) [inline bundle.go]: Remove The

File: README.md [hunk 12]
```diff
@@ -14,6 +14,7 @@ VERSION:
    
 COMMANDS:
    generate     generate a OCI spec file
+   bundleValidate verify if a bundle was valid
```
Comment (mrunalp) [inline README.md]: s/was/is

**Comment by mrunalp** [comment]
Thanks! I will test this out today.

File: bundle.go [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+		}
+		defer rf.Close()
+
+		var runtime specs.LinuxRuntimeSpec
+		if err = json.NewDecoder(rf).Decode(&runtime); err != nil {
+			logrus.Fatal(err)
+		}
+
+		bundleValidate(spec, runtime, path.Join(inputPath, "rootfs"))
+		logrus.Infof("Bundle test succeeded.")
```
Comment (mrunalp) [inline bundle.go:69]: s/test/validation

File: bundle.go [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+			logrus.Fatal(err)
+		}
+		defer rf.Close()
+
+		var runtime specs.LinuxRuntimeSpec
+		if err = json.NewDecoder(rf).Decode(&runtime); err != nil {
+			logrus.Fatal(err)
+		}
+
+		bundleValidate(spec, runtime, path.Join(inputPath, "rootfs"))
```
Comment (mrunalp) [inline bundle.go]: This cannot be assumed. This should be read from root/path in config.json

**Comment by mrunalp** [comment]
Also, I tried testing this with rootfs.tar.gz in this repo and it doesn't work. Could you modify the rootfs so that bundleValidate works with it?

**Comment by wking** [comment]
Re: updating rootfs.tar.gz, see #14.

**Comment by liangchenye** [comment]
close and PR to opencontainers/ocitools

**Pull request closed without merging**